resource-validator/src/main/scala/com/broadinstitute/dsp/resourceValidator/StoppedRuntimeChecker.scala

```diff
@@ -0,0 +1,106 @@
+                  // In contrast to in Leo, we're not setting the shutdown script metadata before stopping the instance
+                  // in order to keep things simple since our main goal here is to prevent unintended cost to users.
+                  deps.computeService
+                    .stopInstance(runtime.googleProject, zoneName, InstanceName(runtime.runtimeName))
+                    .void
+                    .as(Option(runtime))
+            else F.pure(none[Runtime])
+          }
+        } yield runningRuntimeOpt
+
+      private def checkDataprocCluster(runtime: Runtime, isDryRun: Boolean)(
+        implicit ev: Ask[F, TraceId]
+      ): F[Option[Runtime]] = {
+        val clusterName = DataprocClusterName(runtime.runtimeName)
+        val project = runtime.googleProject
+
+        for {
+          clusterOpt <- deps.dataprocService
+            .getCluster(runtime.googleProject, regionName, DataprocClusterName(runtime.runtimeName))
+          runningClusterOpt <- clusterOpt.flatTraverse { cluster =>
+            if (cluster.getStatus.getState.name() == "RUNNING") {
+              for {
+                runningInstanceExists <- containsRunningInstance(project, zoneName, regionName, clusterName)
+                rt <- if (runningInstanceExists) {
+                  if (isDryRun)
+                    logger.warn(s"Cluster (${runtime}) is running. It needs to be stopped.").as(Option(runtime))
+                  else
+                    logger.warn(s"Cluster (${runtime}) is running. Going to stop it.") >> deps.dataprocService
+                    // In contrast to in Leo, we're not setting the shutdown script metadata before stopping the instance
+                    // in order to keep things simple since our main goal here is to prevent unintended cost to users.
+                      .stopCluster(project, regionName, clusterName, metadata = None)
+                      .void
+                      .as(Option(runtime))
+                } else F.pure(none[Runtime])
+              } yield rt
+            } else F.pure(none[Runtime])
+          }
+        } yield runningClusterOpt
+      }
+
+      private def containsRunningInstance(project: GoogleProject,
+                                          zone: ZoneName,
+                                          region: RegionName,
+                                          cluster: DataprocClusterName): F[Boolean] =
+        for {
+          instances <- deps.dataprocService.getClusterInstances(project, region, cluster)
+          doesThereExistARunningInstance <- Stream
+            .emits(instances.values.toSeq.flatten)
+            .covary[F]
+            .parEvalMapUnordered(10) { instance =>
+              deps.computeService
+                .getInstance(project, zone, instance)
+                .handleErrorWith(_ => F.pure(None))
+            }
+            .exists(_.exists(_.getStatus == "RUNNING"))
+            .compile
+            .lastOrError
```

Comment:
Is there a better way to go from `Stream[F, Boolean]` to `F[Boolean]` in this case?

Reply:
`lastOrError` is the right one to use

Reply:
thanks